core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -51,8 +185,744 @@ trait DataSkippingReader
+    //
+    // constructDataFilters(NOT(AND(a, b)))
+    // ==> constructDataFilters(OR(NOT(a), NOT(b)))
+    // ==> OR(constructDataFilters(NOT(a)), constructDataFilters(NOT(b)))
+    //
+    // Assuming we can push the resulting NOT operations all the way down to some leaf operation it
+    // can fold into, the rewrite allows us to create a data skipping filter from the expression.
+    //
+    // a b AND(a, b)
+    // | | | NOT(AND(a, b))
+    // | | | | OR(NOT(a), NOT(b))
+    // T T T F F
+    // T F F T T
+    // T N N N N
+    // F F F T T
+    // F N F T T
+    // N N N N N
+    case Not(And(e1, e2)) =>
+      constructDataFilters(Or(Not(e1), Not(e2)))
+
+    // Push skipping predicate generation through OR (similar to AND case).
+    //
+    // constructDataFilters(OR(a, b))
+    // ==> OR(constructDataFilters(a), constructDataFilters(b))
+    //
+    // Similar to AND case, if the rewritten predicate does not evaluate to TRUE, then it means that
+    // neither `constructDataFilters(a)` nor `constructDataFilters(b)` evaluated to TRUE, which in
+    // turn means that neither `a` nor `b` could evaulate to TRUE for any row the file might
+    // contain, which proves we have a valid data skipping predicate.
+    //
+    // Unlike AND, a single leg of an OR expression provides no filtering power -- we can only
+    // reject a file if both legs evaluate to false.
+    case Or(e1, e2) =>
+      val e1Filter = constructDataFilters(e1)
+      val e2Filter = constructDataFilters(e2)
+      if (e1Filter.isDefined && e2Filter.isDefined) {
+        Some(DataSkippingPredicate(
+          e1Filter.get.expr || e2Filter.get.expr,
+          e1Filter.get.referencedStats ++ e2Filter.get.referencedStats))
+      } else {
+        None
+      }
+
+    // Similar to AND, we can (and want to) push the NOT past the OR using deMorgan's law.
+    case Not(Or(e1, e2)) =>
+      constructDataFilters(And(Not(e1), Not(e2)))
+
+    // Match any file whose null count is larger than zero.
+    case IsNull(SkippingEligibleColumn(a)) =>
+      val nullCountCol = StatsColumn(NULL_COUNT, a)
+      getStatsColumnOpt(nullCountCol).map { nullCount =>
+        DataSkippingPredicate(nullCount > Literal(0), nullCountCol)
+      }
+    case Not(IsNull(e)) =>
+      constructDataFilters(IsNotNull(e))
+
+    // Match any file whose null count is less than the row count.
+    case IsNotNull(SkippingEligibleColumn(a)) =>
+      val nullCountCol = StatsColumn(NULL_COUNT, a)
+      val numRecordsCol = StatsColumn(NUM_RECORDS)
+      getStatsColumnOpt(nullCountCol).flatMap { nullCount =>
+        getStatsColumnOpt(numRecordsCol).map { numRecords =>
+          DataSkippingPredicate(nullCount < numRecords, nullCountCol, numRecordsCol)
+        }
+      }
+    case Not(IsNotNull(e)) =>
+      constructDataFilters(IsNull(e))
+
+    // Match any file whose min/max range contains the requested point.
+    case EqualTo(SkippingEligibleColumn(a), SkippingEligibleLiteral(v)) =>
```

Comment:
Shall we also handle the expression `EqualNullSafe`?

Reply:
Another good suggestion. Much like the comment above, once this PR is merged, want to add this change yourself?

Reply:
will raise a PR